adr/0017-agent-operations-framework.md

```diff
@@ -0,0 +1,129 @@
+# 17. Agent Operations Framework
+
+Date: 2025-01-27
+
+## Status
+
+Accepted
+
+## Context
+
+Following the new epics and product decisions, we need to build a framework capable of executing write operations on SAP machines.
+To achieve this, we have decided to use the agent, as it runs on the machines with root privileges, has all the necessary permissions to execute commands, and includes infrastructure to consume messages from other components of Trento.
+
+Unlike fact-gathering operations, we have chosen to delegate this new development to a dedicated library rather than embedding the code directly into the agent's codebase. This approach ensures better separation of concerns and allows the release of additional tools supporting operations without affecting the agent's codebase.
+
+The requirements for this new development are very specific:
+
+- Operations must be atomic and include rollback capabilities.
+- Operations must accept arguments to enable actions in different contexts without requiring dedicated code for each case.
```

Comment:
question: I believe I am missing some details here
> Operations must accept arguments to enable actions in different contexts without requiring dedicated code for each case
>
Could you help out with an example or point to some extra information to properly understand, please?

Reply:
The first operation we are working on is `saptune`. Applying saptune solutions specifically.
Different hosts need different solutions. We have `HANA`, `NETWEAVER`, `S4HANA`, `HANA=S4HANA`, etc.
So we are going to send this value as argument. This way, the operation will do a slightly different thing based on the argument.